Documentation/contributing/development/statedb.rst

```diff
@@ -0,0 +1,443 @@
+
+  contrib/examples/statedb$ go run .
+
+
+Pitfalls
+^^^^^^^^
+
+Here are some common mistakes to be aware of:
+
+* Object is mutated after insertion to database. Since StateDB queries do not return copies, all readers will see the modifications.
+
+* Object (stored by reference, e.g. ``*T``) returned from a query is mutated and then inserted. StateDB will catch this and panic. Objects stored by reference must be (shallow) cloned before mutating.
+
+* Query is made with ReadTxn and results are used in a WriteTxn. The results may have changed between the ReadTxn and WriteTxn! If you want optimistic concurrency control, then use CompareAndSwap in the write transaction.
+
+cilium-dbg statedb
+^^^^^^^^^^^^^^^^^^
+
+StateDB implements a simple HTTP API for querying tables from a separate process.
+The objects are marshalled and unmarshalled using the normal Go JSON facilities,
+so we can effectively get copies of the objects into a separate process using the
+HTTP API which we can then render or process.
+
+The ``cilium-dbg statedb <command>`` provides a table-based view into the StateDB
+tables running in the agent in a style similar to "kubectl get". Watching the
+tables for changes is also supported with the ``--watch`` flag.
+
+To get your newly defined table to be part of this, there's only two things you
+need to do: implement the *TableRow* and *TableHeader* methods for the object (see above)
+and add a one-liner to ``cilium-dbg/cmd/statedb.go``:
+
+.. code-block:: go
+
+	func init() {
+		...
+		StatedbCmd.AddCommand(
+			...
+			statedbTableCommand[example.Example](example.TableName),
+		)
+	}
+
+We can now dump the contents of our example table:
+
+.. code-block:: shell
+
+	$ cilium-dbg statedb examples
+	# ID      Label
+	123       hello  
+
+Kubernetes reflection
+^^^^^^^^^^^^^^^^^^^^^
+
+To reflect Kubernetes objects from the API server into a table, the reflector
+utility in ``pkg/k8s`` can be used to automate this. For example, we can define
+a table of pods and reflect them from Kubernetes into the table:
+
+.. literalinclude:: ../../../contrib/examples/statedb_k8s/pods.go
+   :language: go
+
+As earlier, we can then construct a small application to try this out:
+
+.. literalinclude:: ../../../contrib/examples/statedb_k8s/main.go
+   :language: go
+
+You can run the example in ``contrib/examples/statedb_k8s`` to watch the pods in
+your current cluster:
+
+.. code-block:: shell
+
+    contrib/examples/statedb_k8s$ go run . --k8s-kubeconfig-path ~/.kube/config
+    level=info msg=Starting
+    time="2024-09-05T11:22:15+02:00" level=info msg="Establishing connection to apiserver" host="https://127.0.0.1:44261" subsys=k8s-client
+    time="2024-09-05T11:22:15+02:00" level=info msg="Connected to apiserver" subsys=k8s-client
+    level=info msg=Started duration=9.675917ms
+    Pod(default/nginx): Running (revision: 1, deleted: false)
+    Pod(kube-system/cilium-envoy-8xwp7): Running (revision: 2, deleted: false)
+    ...
+
+Reconcilers
```

Comment:
Should we add some details around the difference of using table pointers in context of reconcilers?